ote/src/clj/ote/transit_changes/detection.clj

```diff
@@ -470,23 +474,39 @@
-                    {:gtfs/transport-service-id service-id
-
-                     :gtfs/date today
-                     :gtfs/change-date new-change-date
-                     :gtfs/different-week-date (:gtfs/different-week-date earliest-route-change)
-                     :gtfs/current-week-date (:gtfs/current-week-date earliest-route-change)
-
-                     :gtfs/removed-routes (:removed change-count-by-type 0)
-                     :gtfs/added-routes (:added change-count-by-type 0)
-                     :gtfs/changed-routes (:changed change-count-by-type 0)
-                     :gtfs/no-traffic-routes (:no-traffic change-count-by-type 0)
-                     :gtfs/route-changes route-change-infos
-
-                     :gtfs/package-ids package-ids
-                     :gtfs/created (java.util.Date.)})))
+    (tx/with-transaction db
+                         (specql/upsert! db :gtfs/transit-changes
+                                         #{:gtfs/transport-service-id :gtfs/date}
+                                         {:gtfs/transport-service-id service-id
+
+                                          :gtfs/date today
+                                          :gtfs/change-date new-change-date
+                                          :gtfs/different-week-date (:gtfs/different-week-date earliest-route-change)
+                                          :gtfs/current-week-date (:gtfs/current-week-date earliest-route-change)
+
+                                          :gtfs/removed-routes (:removed change-count-by-type 0)
+                                          :gtfs/added-routes (:added change-count-by-type 0)
+                                          :gtfs/changed-routes (:changed change-count-by-type 0)
+                                          :gtfs/no-traffic-routes (:no-traffic change-count-by-type 0)
+
+                                          :gtfs/package-ids package-ids
+                                          :gtfs/created (java.util.Date.)})
+                         (doseq [r route-change-infos]
+                           (let [update-count (specql/update! db :gtfs/detected-route-change
+                                                              (merge {:gtfs/transit-change-date (sql-date today)
+                                                                      :gtfs/transit-service-id service-id
+                                                                      :gtfs/created-date (java.util.Date.)}
+                                                                     r)
+                                                              {:gtfs/transit-change-date (sql-date today)
+                                                               :gtfs/transit-service-id service-id
+                                                               :gtfs/route-hash-id (:gtfs/route-hash-id r)})]
+
+                             (when (not= 1 update-count)
```

Comment:
for future reference, specql also supports upsert - but no need to change now as the code is tested & working.

Reply:
There was a reason why this was done as update&insert pair...some problem when utilizing upsert.